src/Units/UnitState.java

```diff
@@ -82,7 +82,7 @@ public void fire(WeaponModel weapon)
 
   public boolean hasMaterials()
   {
-    return materials > 0 || materials == model.maxMaterials;
+    return materials > 0 || !model.needsMaterials;
```

Comment:
If a unit has no materials, and also doesn't need materials, then it `hasMaterials()`?

Reply:
Should it be `canBuild()`, then? That requires a little bit of logic to take you to "this cares about materials", and doesn't necessarily account for other uses we may have down the line.
Could call it `canUseMaterials()`, but that's roughly equivalent and not as nice to say.

Reply:
I think I'd take `canBuild()` over `canUseMaterials()` since this function is also checking whether materials are present, not just whether they would be usable.

Or just change the logic so that it returns `false` if `model.needsMaterials` is `false`, which you'd have to do anyway for `canBuild()` to make sense as the name.

Reply:
Eh? `needsMaterials = false` means `canBuild()` should always return true.
If I don't need materials, then I can build regardless of whether I have any.

Reply:
Do we have any units that produce things without materials? I tried to explore how materials are actually handled, and all I found is that carriers are broken on the KaijuOnly branch: `WARNING! UnitProduction Action created with invalid arguments.`

Reply:
Oh, goody. I guess I should make a test on that, then :/
...OTOH, it's not like we'll be touching this code again. What possibilities exist aside from "no materials", "finite materials", and "unlimited materials"?

I decided the Sky Carrier shouldn't have a materials limit (since there aren't any other limits like that in its source game), which is why I added this behavior.

Reply:
Uhhh, I dunno where you're getting that Carriers are broken.
I just tested on both `kaijuOnly` and `kaijuExtras`, and DoR Carriers build stuff just fine.

Also also, that error message only comes from trying to build from a property (or from `UnitSpawnAction`... I guess I should fix that)

Reply:
Turns out I was just trying to build it with insufficient funds. Really need better visuals for that. Silly me.

As for possibilities; I think that covers the bases (though: what's the difference between none and infinity?). Also, we should probably show the current materials count in the unit preview overlay (unless it's infinite).

Reply:
None means you can't build anything, though I guess that's not really a different *case* from finite materials - it's just that the finite quantity is currently 0. Blame tired brains.
Infinity means you can build as much as you like.

Also, we totally have an issue for that already: #275